### ISSUE TITLE
Performance regression: epoch exceeds slow-runtime threshold

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#454]
During training, the one-epoch wall time remains active after 60 seconds, exceeding the slow-runtime threshold. The observed cause is a performance bottleneck in the training loop, likely due to the current backend nv0 (likely a CPU or slower GPU implementation). The configuration includes a model with 8 perceptrons, pooling 2, loss MSE, and training for one epoch with adamw optimizer. This indicates the backend cannot process the model data and operations within expected throughput.

State: The failure was observed with the provided reproduction packet, which uses the configured slow-runtime threshold as the boundary condition. The backend nv0 currently fails to complete a single training epoch within the threshold, suggesting the implementation may not be optimized for this workload.

Occurrence: The failure occurs at the epoch boundary when the training loop is evaluating the loss and updating parameters, resulting in the runtime exceeding the configured threshold.

Next steps: The issue requires benchmarking and optimization of the backend nv0 or replacing with a faster implementation to meet performance expectations for one training epoch.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `424c87997fb30ec1`
- Classifier provider: `openrouter`
- Classifier model: `cohere/north-mini-code:free`
- Classifier effort: `default`

### Classifier decision

```json
{
  "verdict": "new",
  "issue": 0,
  "title": "Performance regression: epoch exceeds slow-runtime threshold",
  "body": "During training, the one-epoch wall time remains active after 60 seconds, exceeding the slow-runtime threshold. The observed cause is a performance bottleneck in the training loop, likely due to the current backend nv0 (likely a CPU or slower GPU implementation). The configuration includes a model with 8 perceptrons, pooling 2, loss MSE, and training for one epoch with adamw optimizer. This indicates the backend cannot process the model data and operations within expected throughput.\n\nState: The failure was observed with the provided reproduction packet, which uses the configured slow-runtime threshold as the boundary condition. The backend nv0 currently fails to complete a single training epoch within the threshold, suggesting the implementation may not be optimized for this workload.\n\nOccurrence: The failure occurs at the epoch boundary when the training loop is evaluating the loss and updating parameters, resulting in the runtime exceeding the configured threshold.\n\nNext steps: The issue requires benchmarking and optimization of the backend nv0 or replacing with a faster implementation to meet performance expectations for one training epoch.",
  "rationale": "No existing GitHub issues address performance bottlenecks that cause epoch wall times to exceed 60 seconds for this configuration, and the current search of performance-related issues yielded no matches. The evidence indicates a new performance regression in the backend implementation that needs optimization.",
  "error_type": "performance_regression"
}
```

## Deterministic failure packet

kind=performance
id=424c87997fb30ec1
base=commit=9cc65a63b4677c53372379219cb280da9085ac1b tracked_tree=clean
cursor=seed:0 cursor:17194 next:17195 composition:54772174239326
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:nv0
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-machine-workload/target/debug/recipe-trials/recipe-trial-1260-17194/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/chronological_splits")
		.target("target")
		.split(0.8);
	let model = recipe.model()
		.perc(8).elu().qi(4).1
		.pool(2).gelu().norm(batch).iq(2).m
		.loss(mse);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(54772174239326)
		.epochs(1)
		.log(all)
		.stop(0.8)
		.int(1)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv0

**Comment by nm-z** [comment]
The training epoch exceeds the slow‑runtime threshold on the nv1 backend with the given model composition, matching existing issue 454.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `298d2a5cd48477c7`
- Classifier provider: `agy`
- Classifier model: `gpt-oss-120b-medium`
- Classifier effort: `medium`

### Classifier decision

```json
{"body":"The training epoch exceeds the slow‑runtime threshold on the nv1 backend with the given model composition, matching existing issue 454.","issue":454,"rationale":"Search of issues.json found issue 454 covering epoch exceeding the slow‑runtime threshold. The failure occurs at the earliest public operation recipe.train().run() with the same implementation path, so it maps to the existing issue.","title":"Performance regression: epoch exceeds slow-runtime threshold","verdict":"comment"}
```

## Deterministic failure packet

kind=performance
id=298d2a5cd48477c7
base=commit=95680a595baee3ff942a7fe4f5f19d2f17a08448 tracked_tree=clean
cursor=seed:0 cursor:15768 next:15769 composition:50116574142792
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:nv1
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-issue-machine/target/debug/recipe-trials/recipe-trial-18-15768/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/chronological_splits")
		.target("target");
	let model = recipe.model()
		.rnn(8).log().iq(4).xs
		.perc(8).sigmoid().qi(4).nf
		.loss(mae);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(50116574142792)
		.epochs(1)
		.log(all)
		.stop(0.0)
		.f(6, 9)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv1

**Comment by nm-z** [comment]
Closing as resolved based on zero-change verification PR #524.

That PR changes 0 files and verifies the existing `minimal` behavior, so there is no code to merge. The original verification is copied below.

<details>
<summary>Verification from PR #524: Verify epoch runtime stays below the threshold</summary>

## Summary

- Reproduce both deterministic performance packets on current minimal.
- Confirm that neither public training path approaches the 60-second threshold.

## Verification

- `nv0`: the training epoch took 42.365 ms, evaluation took 3.0925 seconds, and the complete public run took 3.836 seconds.
- `nv1`: the training epoch took 550.026 ms, and the complete public run took 5.340 seconds.
- Public inference loaded both saved bundles and returned finite outputs `0` and `24.5` in 2.663 seconds.
- Archy finished with 22 GiB available, zero swap use, and all six GPUs idle at 26 to 36°C.

Fixes #454

</details>

Accepted verdict: PASS

Reason: The verification above reports that current `minimal` completes this issue's reported path successfully.
